### PR TITLE
SoundBank

### PR DESCRIPTION
**Pull request by gnarf**
(just up for ease of communication, this is purely a first draft right now)

### CODE REVIEW

File: src/SoundBank.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        this.playerTargets.forEach((playerTarget, key) => {
+            if (playerTarget === target) {
+                this.getSoundEffects(key).setFromTarget(target);
+            }
+        });
+    }
+
+    stop (target, soundId) {
+        if (this.playerTargets.get(soundId) === target) {
+            this.soundPlayers[soundId].stop();
```
Comment (gnarf) [inline src/SoundBank.js]: should probaby use `this.getSound(soundId)` here for future refactor possibilities.

File: src/SoundBank.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        });
+    }
+
+    dispose () {
+        this.playerTargets.clear();
+        this.soundEffects.forEach(effects => effects.dispose());
+        this.soundEffects.clear();
+        for (const soundId in this.soundPlayers) {
+            if (this.soundPlayers.hasOwnProperty(soundId)) {
+                this.soundPlayers[soundId].dispose();
```
Comment (gnarf) [inline src/SoundBank.js]: I feel like this one can stay as is in the dispose loop, no need to use getSound

File: src/SoundBank.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        this.soundPlayers = {};
+        this.playerTargets = new Map();
+        this.soundEffects = new Map();
+    }
+
+    getSound (soundId) {
+        if (!this.soundPlayers[soundId]) {
+            this.soundPlayers[soundId] = new SoundPlayer(this.audioEngine, {
+                id: soundId, buffer: this.audioEngine.audioBuffers[soundId]
```
Comment (mzgoddard) [inline src/SoundBank.js]: We should get rid of `audioBuffers`. Its a memory leak waiting to happen.

File: src/SoundBank.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
+const SoundPlayer = require('./GreenPlayer');
+const EffectsChain = require('./EffectsChain');
+
+const ALL_TARGETS = '*';
+
+class SoundBank {
+    constructor (audioEngine) {
+        this.audioEngine = audioEngine;
```
Comment (mzgoddard) [inline src/SoundBank.js]: I think it'd be best if SoundBank does not know of AudioEngine.

File: src/effects/EffectChain.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            return effect;
+        }, this.audioEngine);
+
+        this._soundPlayers = new Set();
+    }
+
+    addSoundPlayer (soundPlayer) {
+        if (!this._soundPlayers.has(soundPlayer)) {
+            this._soundPlayers.add(soundPlayer);
+            this._effects.forEach(effect => effect.update());
```
Comment (mzgoddard) [inline src/effects/EffectChain.js]: I think making update a function to mirror Effect's API would be good.

File: src/effects/EffectChain.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+class EffectChain {
+    constructor (audioEngine) {
```
Comment (mzgoddard) [inline src/effects/EffectChain.js]: Should we have an `effects` argument?

File: src/effects/EffectChain.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    setEffectsFromTarget (target) {
+        this._effects.forEach(effect => {
+            if (effect.name in target) {
+                effect.set(target[effect.name]);
+            } else if ('soundEffects' in target && effect.name in target.soundEffects) {
+                effect.set(target.soundEffects[effect.name]);
+            } else {
+                effect.set(effect.DEFAULT_VALUE);
```
Comment (mzgoddard) [inline src/effects/EffectChain.js]: This feels iffy.

File: src/effects/EffectChain.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+            this.lastEffect = effect;
+            return effect;
+        });
+
+        // Walk backwards through effects connecting the last output to audio engine,
+        // then each effect's output to the input of the next effect.
+        this._effects.reduceRight((nextNode, effect) => {
+            effect.connect(nextNode);
+            return effect;
+        }, this.audioEngine);
```
Comment (mzgoddard) [inline src/effects/EffectChain.js]: Lets make connecting to audioEngine done through a `.connect` method like Effect and SoundPlayer.

File: src/SoundBank.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+                id: soundId, buffer: this.audioEngine.audioBuffers[soundId]
+            });
+        }
+
+        return this.soundPlayers[soundId];
+    }
+
+    getSoundEffects (sound) {
+        if (!this.soundEffects.has(sound)) {
+            this.soundEffects.set(sound, new EffectsChain(this.audioEngine));
```
Comment (mzgoddard) [inline src/SoundBank.js]: I think it'd be better to have a `this.effectsStandard.clone()` way to make new EffectsChains.

File: src/SoundBank.js (hunk 4 above)
Comment (gnarf) [reply src/SoundBank.js]: The only reason it needs it is for `addSound` and knowing the output node to connect to I think...

File: src/effects/EffectChain.js (hunk 7 above)
Comment (gnarf) [reply src/effects/EffectChain.js]: agree - this whole bit didn't work out as nicely as I'd hoped on the VM side

File: src/effects/EffectChain.js (hunk 8 above)
Comment (gnarf) [reply src/effects/EffectChain.js]: you mean that soundbank would connect soundplayer -> effectchain -> audioEngine ?

File: src/SoundBank.js [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    playSound (target, soundId) {
+        const effects = this.getSoundEffects(soundId);
+        const player = this.getSoundPlayer(soundId);
+
+        this.playerTargets.set(soundId, target);
+        effects.setEffectsFromTarget(target);
+        effects.addSoundPlayer(player);
+
+        player.connect(effects);
```
Comment (gnarf) [inline src/SoundBank.js]: need to play before connect

File: src/effects/EffectChain.js [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+            this.inputNode.connect(lastEffect.getInputNode());
+
+            return;
+        } else if (target === firstEffect) {
+            return;
+        }
+
+        this.target = target;
+
+        firstEffect.connect(target);
```
Comment (gnarf) [inline src/effects/EffectChain.js]: is this backwards now?

**Pull request merged**